android/src/main/java/io/grpc/android/AndroidChannelBuilder.java

```diff
@@ -0,0 +1,255 @@
+import io.grpc.MethodDescriptor;
+import io.grpc.internal.GrpcUtil;
+import io.grpc.okhttp.OkHttpChannelBuilder;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Builds a {@link ManagedChannel} that automatically monitors the Android device's network state.
+ * Network changes are propagated to the underlying OkHttp-backed {@ManagedChannel} to smoothly
+ * handle intermittent network failures.
+ *
+ * <p>gRPC Cronet users should use {@code CronetChannelBuilder} directly, as Cronet itself monitors
+ * the device network state.
+ *
+ * <p>Requires the Android ACCESS_NETWORK_STATE permission.
+ *
+ * @since 1.12.0
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/4056")
+public final class AndroidChannelBuilder extends ForwardingChannelBuilder<AndroidChannelBuilder> {
+
+  private final ManagedChannelBuilder delegateBuilder;
+  private final Context context;
+
+  /** Always fails. Call {@link #forAddress(String, int, Context)} instead. */
+  public static AndroidChannelBuilder forTarget(String target) {
+    throw new UnsupportedOperationException("call forTarget(String, Context) instead");
+  }
+
+  /** Always fails. Call {@link #forAddress(String, int, Context)} instead. */
+  public static AndroidChannelBuilder forAddress(String name, int port) {
+    throw new UnsupportedOperationException("call forAddress(String, int, Context) instead");
+  }
+
+  /** Creates a new builder for the given target and Android context. */
+  public static final AndroidChannelBuilder forTarget(String target, Context context) {
+    return new AndroidChannelBuilder(target, context);
+  }
+
+  /** Creates a new builder for the given host, port, and Android context. */
+  public static AndroidChannelBuilder forAddress(String name, int port, Context context) {
+    return forTarget(GrpcUtil.authorityFromHostAndPort(name, port), context);
+  }
+
+  private AndroidChannelBuilder(String target, Context context) {
+    delegateBuilder = OkHttpChannelBuilder.forTarget(target);
+    this.context = context;
+  }
+
+  @Override
+  protected ManagedChannelBuilder<?> delegate() {
+    return delegateBuilder;
+  }
+
+  @Override
+  public ManagedChannel build() {
+    return new AndroidChannel(delegateBuilder.build(), context);
+  }
+
+  /**
+   * Wraps an OkHttp channel and handles invoking the appropriate methods (e.g., {@link
+   * ManagedChannel#resetConnectBackoff}) when the device network state changes.
+   */
+  @VisibleForTesting
+  static final class AndroidChannel extends ManagedChannel {
+
+    private final ManagedChannel delegate;
+    private final Context context;
+    private final ConnectivityManager connectivityManager;
+
+    private DefaultNetworkCallback defaultNetworkCallback;
+    private NetworkReceiver networkReceiver;
+
+    private final Object lock = new Object();
+
+    // May only go from true to false, and lock must be held when assigning this
+    private volatile boolean needToUnregisterListener = true;
```

Comment:
I'd suggest making this just `@GuardedBy("lock")`. It's only going to be rarely used, because shutdown isn't called frequently. Little benefit in optimizing the avoidance of the lock.

Reply:
Done